### DISCUSSION TITLE
Make it clear from a sensor's properties if it represents measurements

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#557] [Entity Models] [answered]
### Proposal:
Make it clear from a sensor's properties if its state represents a present time measurement:
- Classify the type of state: The state could be a *measurement in present time* from a temperature sensor or an energy meter, a *historic value* such as the average temperature during the last 24 hours or the amount of energy used last month, or a *predicted value* such as a weather forecast or the next garbage pickup schedule. Assign a new property `state_class` for this. If `state_class="measurement"`, the state represents a current value, and not a historical aggregation or a prediction of the future. Otherwise, `state_class=None` or `state_class="other"`.

If there is a need in the future, additional state classes may be added to also distinguish between sensors representing aggregated historical data, future predictions, or other types of data.

~~Make it clear from a sensor's properties if it represents a measurement or some calculated or otherwise derived data by distinguishing between~~
~~- What the values represent, e.g. a physical quantity such as temperature or pressure. Assign a new property `measurement_type` for this.~~
~~- What the entity represents, for an entity with `measurement_type=temperature`, this could be CPU temperature, weather forecast, freezer temperature, indoor thermometer, outdoor thermometer.~~
    ~~- With a zigbee temperature sensor as example, the integration does not know if the sensor is placed indoors, outdoors or in the freezer, but it know it's an ambient temperature sensor and not a CPU temperature sensor or a weather forecast sensor. The sensor will then be created as "ambient thermometer", and we allow the user to improve the classification.~~
~~- The source of the data, this could be a measurement from a temperature sensor, statistics such as the average temperature during the last 24 hours, or a calculated value such as weather forecast. Assign a new property `data_source` for this.~~

Initial proposal:
~~- Restricting the current device classes, e.g. `DEVICE_CLASS_TEMPERATURE`, to be used only for measurements and add new device classes, e.g. `DEVICE_CLASS_TEMPERATURE_STATISTIC ` or `DEVICE_CLASS_STATISTIC`~~
~~- Adding a new property to classify the sensor's data as measured or not measured~~

### Background:
Device classes such as `DEVICE_CLASS_TEMPERATURE` can be used for multiple types of sensors, for example:
- A regularly updated temperature measurement
- Historical or statistic data, for example daily average temperature
- Future data, for example tomorrow's forecast

Differentiating between those sensors which represent a measurement and those which don't will be highly useful in order to automatically make a reasonable selection of sensors to include in long-term statistics.

**Comment by balloob** [comment]
[Current device classes documentation](https://developers.home-assistant.io/docs/core/entity/sensor)

I agree. The goal of a device class is that given a state of a certain device class, we know how to interpreted the data. 

Limit the existing sensor device classes to represent the current value except for the following two: 

 - `timestamp`. This is just a timestamp. 
 - `energy`. This already involves a measurement over time so it's kinda meh. We should consider dropping this one in favor of #559 proposal point 3.

**Reply by emontnemery** [reply]
I see two options as outlined above:
- Restricting the current device classes, e.g. DEVICE_CLASS_TEMPERATURE, to be used only for measurements and add new device classes, e.g. DEVICE_CLASS_TEMPERATURE_STATISTIC or DEVICE_CLASS_STATISTIC
- Adding a new property to classify the sensor's data as measured or not measured

The problem with the first one is that it calls for permutations, such as `DEVICE_CLASS_TEMPERATURE_STATISTIC`. I think this means the second will be easier both to use and to extend: Maybe add a new property `data_class` or similar which can be set to `measured`, `calculated`, `aggregated`, `None`?

**Comment by emontnemery** [comment, marked as answer]
This solution is implemented in https://github.com/home-assistant/core/pull/50063:
> Classify the type of state: The state could be a *measurement in present time* from a temperature sensor or an energy meter, a *historic value* such as the average temperature during the last 24 hours or the amount of energy used last month, or a *predicted value* such as a weather forecast or the next garbage pickup schedule. Assign a new property `state_class` for this. If `state_class="measurement"`, the state represents a current value, and not a historical aggregation or a prediction of the future. Otherwise, `state_class=None` or `state_class="other"`.
